storage_controller/src/tenant_shard.rs

```diff
@@ -646,6 +646,49 @@ impl TenantShard {
         Ok(())
     }
 
+    /// Attempt to reschedule the tenant shard to one of its secondary locations.
+    /// If no schedulable secondary is found, roll back any changes to the intent state.
+    /// If the `attached_to` node argument doesn't match the intent state, no changes
+    /// are made and the function returns successfully.
+    pub(crate) fn reschedule_to_secondary(
```

Comment:
I thought about using this for the "node going offline" handling path, but the current behaviour there is that we will reschedule to a non-secondary node if there's a suitable one. I think that behaviour is desirable, but open to being convinced othewise.